### PR TITLE
Fix not to render optional fields when unset

### PR DESCRIPTION
**Pull request by rjalander**
This PR has the changes to fix the issue https://github.com/cdevents/sdk-java/issues/51, not to render the optional fields when null during serialization using an ObjectMapper.

### CODE REVIEW

File: src/test/java/dev/cdevents/CDEventsTest.java
Comment (afrittoli) [inline src/test/java/dev/cdevents/CDEventsTest.java:61]: If I understand the test correctly, what it does it:

- create a CDEvent object, renders it to json
- create a CloudEvent out of the CDEvent
- extract the payload from the CloudEvent and matches is against the CDEvent as json
- compare type and source

This looks good to test check CloudEvent binding, but how does it help to test the unset fields?
Don't we need JSON documents as test data for that?

Comment (rjalander) [reply src/test/java/dev/cdevents/CDEventsTest.java]: Yes, added test data to compare optional fields with the actual CDEvent.
To keep this PR smaller, added the optional field validation only for Core events for now, will create separate PRs to handle tests for other type of events.

File: src/test/java/dev/cdevents/CDEventsTest.java
Comment (afrittoli) [inline src/test/java/dev/cdevents/CDEventsTest.java:342]: I think this can be ok, it guarantees consistent behaviour, however, I'm not 100% it guarantees that optional fields are not included in the rendered JSON.

At this level, I know that:
- if I create an object without optional fields, those optional fields are not set in the object.
- if I render that object without optional fields and parse it back,  those optional fields are not set in the object.
- If a load an object from JSON that does not specify optional fields, those optional fields are not set in the object

I'm still not 100% sure that the rendered JSON will not contain the optional fields set as well, but perhaps that's covered already by other tests - wdyt? 

I see two possible ways we could test that further - possibly not for every event, just once might be enough.
- either read bytes from `inputStream` and compare it with `cdEventJson` converted to bytes. This can be tricky since you may need to compact the json first (to avoid issues with spaces, indentations etc) and also ordering in dicts may be an issue unless the json rendered implementation always renders in the same order
- or add one resource file with the optional fields set to empty value. Load it and check that the optional fields are in the object, with empty or null value.

File: src/test/java/dev/cdevents/CDEventsTest.java
Comment (afrittoli) [inline src/test/java/dev/cdevents/CDEventsTest.java:351]: Do we really need to go through the cloudevent binding in this test?
It does not harm but I thought that would be covered already by previous tests. 
It's fine to keep it if you deem it relevant here.

File: src/main/java/dev/cdevents/models/TaskRunStartedSubject.java
Comment (afrittoli) [inline src/main/java/dev/cdevents/models/TaskRunStartedSubject.java:139]: Shouldn't this be `TaskRun` instead of `PipelineRun`? 
Since this is not caught by tests today, it looks like all cases where we use `equals` exit after the first check, or maybe they return `false` for the wrong reason.

It looks like we should have unit tests for this `equals` function specifically.
I'm a bit surprised that we need this though - this means that every optional field will require an `equals` function with dedicated logic and unit tests, and this adds a lot of complexity too when we will switch to generating the code.

If I read this logic correctly, this is what is implemented (pseudo-code):
```
aIdNoSource = {"id": "a" }
aIdNullSource =  {"id": "a", "source": null}
aIdSource = {"id": "a", "source": ""}
bIdNoSource = {"id": "a" }
bIdNullSource =  {"id": "a", "source": null}
bIdSource = {"id": "a", "source": ""}

aIdNoSource.equals(bIdNoSource)
aIdNullSource.equals(bIdNullSource)
aIdSource.equals(bIdSource)

all other combinations of "equals" are false.
```

I wonder which of these cases is not covered by the language default `equals` implementation?

File: src/test/java/dev/cdevents/CDEventsTest.java
Comment (rjalander) [reply src/test/java/dev/cdevents/CDEventsTest.java:342]: Yes it compares the Subject's content object as whole but not field to filed comparison.
Or I can compare from both the objects that all optional fields are set to null and they are equal, something like;
```
assertThat(cdEvent.getSubject().getContent().getTaskName()).isEqualTo(null);
assertThat(expectedCDEvent.getSubject().getContent().getTaskName()).isEqualTo(null);
```

File: src/test/java/dev/cdevents/CDEventsTest.java
Comment (rjalander) [reply src/test/java/dev/cdevents/CDEventsTest.java:351]: Yes to validate the CDEvent against schema, I need to invoke `CDEvents.cdEventAsCloudEvent(cdEvent);` 

But I think I can move validation part to the `CDEvents.cdEventAsJson(cdEvent);` method itself as `cdEventAsJson()` is invoked from `cdEventAsCloudEvent()` to set the data in the CloudEvent. So that I can remove cloudevent binding from this test also.

File: src/main/java/dev/cdevents/models/TaskRunStartedSubject.java
Comment (rjalander) [reply src/main/java/dev/cdevents/models/TaskRunStartedSubject.java:139]: No, it is the Object of PipelineRun only we are using as per the spec,  https://github.com/cdevents/spec/blob/v0.3.0/core.md#taskrun-started

Actually the first check will return true only If we are passing the same object reference to the `equals` method, but from the test we are passing two different objects one is `expectedCDEvent` object from resource test data and the other created in the test with new object `cdEvent`.

`assertTrue(cdEvent.getSubject().getContent().equals(expectedCDEvent.getSubject().getContent()));
`

And this `equals` method will be covered fully I can debug when running this test itself and returns true as in this test all optional fields are null and all mandatory fields are set to valid and equal when comparing two objects.

Comment (afrittoli) [reply src/main/java/dev/cdevents/models/TaskRunStartedSubject.java:139]: ok, thanks for the clarification.
If I understood it right, Java's default implementation of `equals` will only compare object references, so we need an implementation of `equals` that checks all properties one by one. Hopefully there is a way to generate a standard `equals` implementation in future.

File: src/main/java/dev/cdevents/models/PipelineRunFinishedSubject.java
Comment (afrittoli) [inline src/main/java/dev/cdevents/models/PipelineRunFinishedSubject.java:129]: NIT: Since pipeline name is a `string`, this could be `if (getPipelineName() == that.getPipelineName())` which would return `true` also in case they're both `null`

File: src/test/java/dev/cdevents/CDEventsTest.java
Comment (rjalander) [reply src/test/java/dev/cdevents/CDEventsTest.java]: Now updated the test validation by comparing the JsonNodes created from `inputStream `and `cdEventJson `
This way I can compare actual and expected optional and mandatory fields from the test data.

File: src/main/java/dev/cdevents/models/PipelineRunFinishedSubject.java
Comment (rjalander) [reply src/main/java/dev/cdevents/models/PipelineRunFinishedSubject.java]: Now updated the test validation by comparing the JsonNodes, `equals` method is no longer needed.

File: src/main/java/dev/cdevents/CDEvents.java
Comment (afrittoli) [inline src/main/java/dev/cdevents/CDEvents.java:39]: Thanks - I think this is fine, perhaps in future, we might add a flag to make validation optional.
In the golang SDK I did not include validation in the JSON rendering, I made a `validate` method available instead, so that it is possible to render and invalid CDEvent to JSON, and the producer can decide whether to validate or not.

I did include a call to `validate` in the cloudevents binding to enforce sending of valid events thought (similar to what you had before). 

I wonder whether instead of moving the validation from `cdEventAsCloudEvent` to `cdEventAsJson`, it would be better to add a call to `validate` in the test?

Comment (afrittoli) [reply src/main/java/dev/cdevents/CDEvents.java:39]: I'm also ok with keeping it as it is today.

**Review by afrittoli — APPROVED** [review]
Thanks for the updates!

File: src/main/java/dev/cdevents/CDEvents.java
Comment (rjalander) [reply src/main/java/dev/cdevents/CDEvents.java]: Agree, even this keeps the behavior consistent.
Reverting to keep the validation only for CloudEvents binding, removed for JSON render and added test assertion to validate CDEvent against schema.

**Comment by afrittoli** [comment]
@rjalander is it ok if squash and merge now, or would you like more reviews from other Java SDK maintainers?

**Comment by rjalander** [comment]
I think we are good to merge, If you have no other comments.

**Pull request merged**
